contrib/zenith/inmem_smgr.c

```diff
@@ -41,6 +41,9 @@ inmem_init(void)
 {
 	HASHCTL		hashCtl;
 
+	if (inmem_files && hash_get_num_entries(inmem_files) == 0)
```

Comment:
We could do one of the following:
- make the hash table smaller; 1024 is much larger than is needed for any WAL record,
- switch to simplehash. It's faster and has a reset-function,
- switch to a simple array instead of a hash table. Almost all WAL records touch 1-4 pages, and linear scanning a simple array is faster than a hash table with such small numbers.

Reply:
> Does this make any difference? Every WAL redo call dirties a page, so 'inmem_files' will never be empty.

Oh, I think I get it now. These functions are only used for pages that are evicted from the buffer cache. That should pretty much never happen with a WAL redo function. Replaying a record won't affect more than a handful of pages. XLR_MAX_BLOCK_ID is 32, so not more than that, actually. Are inmem_write() and inmem_read() dead code in practice?

In any case, let's add comments about that!

Reply:
And let's do this:
```
diff --git a/src/backend/tcop/zenith_wal_redo.c b/src/backend/tcop/zenith_wal_redo.c
index 724c4d56bf..3cf4f0ef97 100644
--- a/src/backend/tcop/zenith_wal_redo.c
+++ b/src/backend/tcop/zenith_wal_redo.c
@@ -176,6 +176,8 @@ WalRedoMain(int argc, char *argv[],
 	 */
 	InitializeGUCOptions();
 
+	num_temp_buffers = 64;
+
 	/*
 	 * Parse command-line options.
 	 * TODO
```
The default num_temp_buffers is 1024, which is much more than we need. And DropRelFileNodeAllLocalBuffers() scans through all buffers. In my quick profiling with 'perf', DropRelFileNodeAllLocalBuffers() was taking about 10% of the CPU time (of the WAL redo process), and after changing it to 64, it's not visible in the profile anymore.

Reply:
> > Does this make any difference? Every WAL redo call dirties a page, so 'inmem_files' will never be empty.
> 
> Oh, I think I get it now. These functions are only used for pages that are evicted from the buffer cache. That should pretty much never happen with a WAL redo function. Replaying a record won't affect more than a handful of pages. XLR_MAX_BLOCK_ID is 32, so not more than that, actually. Are inmem_write() and inmem_read() dead code in practice?
> 
> In any case, let's add comments about that!

As far as I remember there are some cases when wal redo writes data directly to file, bypassing shared buffers. This is why this `inmem_file` is needed. Without it (without resetting files) some tests didn't pass.

Reply:
> We could do one of the following:

Actually even 32 (XLR_MAX_BLOCK_ID) pages are not needed because during recovery we restore only target page and ignoring all other.

I will try to check once again what is the maximal number of pages dirtied during wal redo (when it can be larger than 1).

Reply:
> And let's do this:

Makes sense: will do.